src/identity/manager.rs

```diff
@@ -44,18 +52,306 @@ impl fmt::Display for Identity {
 }
 
 #[derive(Clone)]
-pub struct SecretManager {
-    client: CaClient,
+pub struct SecretManager<T: CertificateProvider> {
+    client: T,
+    cache: Arc<RwLock<HashMap<Identity, watch::Receiver<Option<tls::Certs>>>>>,
+}
+
+impl SecretManager<CaClient> {
+    pub fn new(cfg: crate::config::Config) -> SecretManager<CaClient> {
+        let caclient = CaClient::new(cfg.auth);
+        let cache: HashMap<Identity, watch::Receiver<Option<tls::Certs>>> = Default::default();
+        SecretManager {
+            client: caclient,
+            cache: Arc::new(RwLock::new(cache)),
+        }
+    }
 }
 
-impl SecretManager {
-    pub fn new(cfg: crate::config::Config) -> SecretManager {
-        let client = CaClient::new(cfg.auth);
-        SecretManager { client }
+impl<T: CertificateProvider> SecretManager<T> {
+    pub async fn refresh_handler(
+        id: Identity,
+        cache: Arc<RwLock<HashMap<Identity, watch::Receiver<Option<tls::Certs>>>>>,
+        mut ca_client: T,
+        mut sleep_dur: Duration,
+        tx: watch::Sender<Option<tls::Certs>>,
+    ) {
+        info!("refreshing certs for id {} in {:?}", id, sleep_dur);
+        sleep(sleep_dur).await;
+        loop {
+            match ca_client.fetch_certificate(&id).await {
+                Err(e) => {
+                    warn!("Failed cert refresh for id {:?}: {:?}", id, e);
+                    let mut write_locked_cache = cache.write().unwrap();
+                    let Some(certs_rx) = write_locked_cache.get(&id) else {
+                        // Should not be possible, but if there is no receiver
+                        // in the cache, then no one is using these certs that
+                        // are being refreshed, so let's stop refreshing them.
+                        return;
+                    };
+                    if certs_rx.borrow().clone().unwrap().is_expired() {
```

Comment:
Because we do not spawn the refresh handler with None for the cert entry.  We only have None until the initial request has been completed and the cert populated.  After that we spawn the refresh thread.